dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.h

```diff
@@ -0,0 +1,223 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Stopwatch.h>
+#include <Common/UniThreadPool.h>
+#include <Storages/DeltaMerge/LocalIndexerScheduler_fwd.h>
+#include <Storages/KVStore/Types.h>
+#include <Storages/Page/PageDefinesBase.h>
+
+#include <boost/functional/hash.hpp>
+#include <condition_variable>
+
+namespace DB::DM
+{
+
+// Note: this scheduler is global in the TiFlash instance.
+class LocalIndexerScheduler
+{
+public:
+    // The file id of the DMFile.
+    struct DMFileID
+    {
+        explicit DMFileID(PageIdU64 id_)
+            : id(id_)
+        {}
+        PageIdU64 id;
+    };
+    // The page id of the ColumnFileTiny.
+    struct ColumnFileTinyID
+    {
+        explicit ColumnFileTinyID(PageIdU64 id_)
+            : id(id_)
+        {}
+        PageIdU64 id;
+    };
+    using FileID = std::variant<DMFileID, ColumnFileTinyID>;
+
+    struct Task
+    {
+        // Note: The scheduler will try to schedule farely according to keyspace_id and table_id.
+        KeyspaceID keyspace_id;
+        TableID table_id;
+
+        // The file id of the ColumnFileTiny or DMFile.
+        // Used for the scheduler to avoid concurrently adding index for the same file.
+        std::vector<FileID> file_ids;
+
+        // Used for the scheduler to control the maximum requested memory usage.
+        size_t request_memory;
```

Comment:
```suggestion
        const KeyspaceID keyspace_id;
        const TableID table_id;

        // The file id of the ColumnFileTiny or DMFile.
        // Used for the scheduler to avoid concurrently adding index for the same file.
        const std::vector<FileID> file_ids;

        // Used for the scheduler to control the maximum requested memory usage.
        const size_t request_memory;
```